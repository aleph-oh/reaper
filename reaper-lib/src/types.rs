extern crate serde;

use std::rc::Rc;

#[derive(Debug, Eq, PartialEq, Clone)]
pub struct Field {
    pub name: String,
    pub table: String,
}
type Fields = Vec<Field>;

#[derive(Debug, Eq, PartialEq, Clone)]
pub enum ASTNode {
    Select {
        /// NOTE: we should think about making this Rc. We don't want to
        /// deeply clone the entire Vec<Field>.
        fields: Option<Fields>,
        table: Rc<ASTNode>,
        pred: PredNode,
    },
    Join {
        fields: Option<Fields>,
        table1: Rc<ASTNode>,
        table2: Rc<ASTNode>,
        pred: PredNode,
    }, // Note: Rel, Rel can be expressed as select _ from _, _ where True
    Table {
        name: String,
        columns: Vec<String>,
    },
    Concat {
        table1: Rc<ASTNode>,
        table2: Rc<ASTNode>,
    },
}

impl ASTNode {
    pub(crate) fn height(&self) -> usize {
        match self {
            ASTNode::Select {
                fields: _,
                table,
                pred,
            } => table.height().max(pred.height()),
            ASTNode::Join {
                fields: _,
                table1,
                table2,
                pred,
            } => table1.height().max(table2.height()).max(pred.height()),
            ASTNode::Table { name: _, columns: _ } => 1,
            ASTNode::Concat { table1, table2 } => table1.height().max(table2.height()),
        }
    }
}

#[derive(Debug, Eq, PartialEq, Clone)]
pub enum ExprNode {
    Field(Field),
    Int { value: isize },
}

#[derive(Debug, Eq, PartialEq, Clone)]
pub enum PredNode {
    True,
    Lt {
        left: ExprNode,
        right: ExprNode,
    },
    Eq {
        left: ExprNode,
        right: ExprNode,
    },
    And {
        left: Box<PredNode>,
        right: Box<PredNode>,
    },
}

impl ExprNode {
    pub(crate) fn height(&self) -> usize {
        match self {
            ExprNode::Field(_) => 1,
            ExprNode::Int { value: _ } => 1,
        }
    }
}

impl PredNode {
    pub(crate) fn height(&self) -> usize {
        match self {
            PredNode::True => 1,
            PredNode::Lt { left, right } | PredNode::Eq { left, right } => {
                left.height().max(right.height()) + 1
            }
            PredNode::And { left, right } => left.height().max(right.height()) + 1,
        }
    }
}

#[derive(Debug, Eq, PartialEq, Hash, Clone, serde::Deserialize)]
pub struct ConcTable {
    pub name: String,
    pub columns: Vec<String>,
    pub values: Vec<Vec<isize>>,
}

pub type Example = (Vec<ConcTable>, ConcTable);
