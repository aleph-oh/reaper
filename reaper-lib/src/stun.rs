use bitvec::prelude as bv;
use core::num;
use std::sync::atomic::{AtomicBool, Ordering};
use thiserror::Error;

use crate::types::*;

#[derive(Copy, Clone, Error, Debug)]
pub enum InvalidQueryError {
    #[error("too few predicates, expected 1, got 0")]
    TooFewPredicates,
    #[error("too many predicates, expected 1, got `{0}`")]
    TooManyPredicates(usize),
}

pub fn synthesize_pred<'a>(
    queries: impl Iterator<Item = &'a ASTNode>,
    examples: Examples,
) -> Result<Vec<PredNode>, Vec<(&'a ASTNode, InvalidQueryError)>> {
    // TODO: are we handling the zero-predicate case well?
    let (queries, errors): (Vec<_>, Vec<_>) = queries
        .map(|query| (query, AbstractQuery::try_from(query)))
        .partition(|(_q, r)| Result::is_ok(r));
    if errors.len() > 0 {
        return Err(errors
            .into_iter()
            .map(|(q, r)| (q, r.unwrap_err()))
            .collect());
    }

    let mut predicates: Vec<_> = queries
        .into_iter()
        .filter_map(|(_q, r)| synthesize(&r.unwrap(), &examples))
        .collect();
    predicates.sort_unstable_by_key(PredNode::height);
    Ok(predicates)
}

impl ExprNode {
    fn height(&self) -> usize {
        match self {
            ExprNode::FieldName { name: _ } => 1,
            ExprNode::Int { value: _ } => 1,
        }
    }
}

impl PredNode {
    fn height(&self) -> usize {
        match self {
            PredNode::Lt { left, right } | PredNode::Eq { left, right } => {
                left.height().max(right.height()) + 1
            }
            PredNode::And { left, right } => left.height().max(right.height()) + 1,
        }
    }
}

/// An [IntermediateTable] references a ConcreteTable and contains
/// a bit-vector of size [t.values.len()] where bit i is set if
/// row i is in the table.
struct IntermediateTable<'a> {
    base: &'a ConcTable,
    rows: bv::BitVec,
}

impl<'a> IntermediateTable<'a> {
    /// [new(t)] returns a new concrete table which represents the same abstract state as [t].
    fn new(base: &'a ConcTable) -> Self {
        let rows = bv::bitvec![1; base.values.len()];
        IntermediateTable { base, rows }
    }

    fn columns(&self) -> &[String] {
        // NOTE: I don't like this return type, but &str and String
        // have a different memory layout so there's no non-allocating
        // way to transform the type.
        self.base.columns.as_ref()
    }
}

// NOTE: if &= ends up being useful, figure out how to implement it. The overloads for
// BitVec don't seem that helpful.

/// [run_unless_stopped] runs [f] unless [stopper] is set.
///
/// It's recommended that [f] itself uses [run_unless_stopped] during different phases of
/// expensive computation to ensure that the request to stop is respected as best as possible.
fn run_unless_stopped<T>(f: impl FnOnce() -> T, stopper: &AtomicBool) -> Option<T> {
    match stopper.load(Ordering::SeqCst) {
        true => None,
        false => Some(f()),
    }
}

impl ASTNode {
    fn num_holes(&self) -> usize {
        match self {
            ASTNode::Select { table, .. } => table.num_holes() + 1,
            ASTNode::Join { table1, table2, .. } => table1.num_holes() + table2.num_holes() + 1,
            ASTNode::Table { .. } => 0,
            ASTNode::Field { .. } => 0,
        }
    }
}

#[derive(Debug)]
struct AbstractQuery(ASTNode);

impl TryFrom<&ASTNode> for AbstractQuery {
    type Error = InvalidQueryError;

    fn try_from(value: &ASTNode) -> Result<Self, Self::Error> {
        match value.num_holes() {
            0 => Err(InvalidQueryError::TooFewPredicates),
            1 => Ok(AbstractQuery(value.clone())),
            n => Err(InvalidQueryError::TooManyPredicates(n)),
        }
    }
}

fn synthesize(query: &AbstractQuery, examples: Examples) -> Option<PredNode> {
    // The Scythe paper implements predicate search as a top-down search where we try to generate predicates (simplest-first)
    // such that they produce the expected output. It is essentially an exhaustive search. This leaves a few questions:
    //  - How do we group predicates? We probably want to map a given query result to all the predicates that produce
    //  that query result.
    //      - How do we do this pruning before concrete evaluation to reduce the number of concrete evaluations?
    //  - This predicate grouping also doesn't handle the case where a query returns too many entries.
    // TODO: how do we limit the amount of time the synthesizer spends?
    //  - limit the depth and the time together, limiting time is a little
    //  less invasive
    // TODO: how do we parallelize this nicely?
    //  - rayon, spawns / parallel iteration have a nice / reasonable API
    // TODO: should we be only returning one candidate predicate? Maybe we can
    // return many if time allows. Combining this with parallelism will require some kind of concurrency-safe vector
    // (or we can put a Mutex around the Vec).
    todo!()
}
