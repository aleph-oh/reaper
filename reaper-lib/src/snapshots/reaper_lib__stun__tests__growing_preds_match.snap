---
source: src/stun.rs
expression: grown
---
[
    And {
        left: Lt {
            left: Int {
                value: 1,
            },
            right: Int {
                value: 2,
            },
        },
        right: Lt {
            left: Int {
                value: 1,
            },
            right: Int {
                value: 2,
            },
        },
    },
    And {
        left: Lt {
            left: Int {
                value: 1,
            },
            right: Int {
                value: 2,
            },
        },
        right: Lt {
            left: FieldName {
                name: "hello",
            },
            right: FieldName {
                name: "world",
            },
        },
    },
    And {
        left: Lt {
            left: Int {
                value: 1,
            },
            right: Int {
                value: 2,
            },
        },
        right: True,
    },
    And {
        left: Lt {
            left: FieldName {
                name: "hello",
            },
            right: FieldName {
                name: "world",
            },
        },
        right: Lt {
            left: Int {
                value: 1,
            },
            right: Int {
                value: 2,
            },
        },
    },
    And {
        left: Lt {
            left: FieldName {
                name: "hello",
            },
            right: FieldName {
                name: "world",
            },
        },
        right: Lt {
            left: FieldName {
                name: "hello",
            },
            right: FieldName {
                name: "world",
            },
        },
    },
    And {
        left: Lt {
            left: FieldName {
                name: "hello",
            },
            right: FieldName {
                name: "world",
            },
        },
        right: True,
    },
]
